### ISSUE TITLE
Spell-checking of search terms

### ISSUE THREAD

**Issue by zackkrida** [WordPress/openverse-api#701]
## Problem

If someone makes a typo with an API request, they might not realize and could get undesirable results.

## Description

We should implement spell-checking in the api. Perhaps the response could just have a new field for a `suggested_term` or some other method of letting the user know that there's a typo in their query. I've never implemented anything like this before and am not sure of the best approaches to take!

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

**Comment by dhruvkb** [comment]
This might not be 100% directly related but we do have a list of the [3000 most commonly used words in English](https://github.com/WordPress/openverse-api/blob/6cc0551c639081f895d974b30cc465713d0ce360/api/test/load_testing/api/common_english_words.txt)  in the API. Making spell-check work with i18n might be an interesting challenge.

**Comment by sarayourfriend** [comment]
Spell checking with i18n would be difficult, but [hunspell](https://hunspell.github.io/) and similar libraries could be helpful here. If we wanted a service that did this for us instead of calling out to a library (for some reason, I'm not sure it's worth the extra complexity, but who knows) there's also https://languagetool.org/dev that we could run for ourselves.

It's pretty powerful. Here's a request for `burds` with the language set to `en-US`.

```
curl -X POST --header 'Content-Type: application/x-www-form-urlencoded' --header 'Accept: application/json' -d 'text=burds&language=en-US&enabledOnly=false' 'https://api.languagetoolplus.com/v2/check'
```

<details>
<summary>JSON response blob</summary>
<pre>
{
  "software": {
    "name": "LanguageTool",
    "version": "5.8-SNAPSHOT",
    "buildDate": "2022-05-11 08:08:33 +0000",
    "apiVersion": 1,
    "premium": true,
    "premiumHint": "You might be missing errors only the Premium version can find. Contact us at support<at>languagetoolplus.com.",
    "status": ""
  },
  "warnings": {
    "incompleteResults": false
  },
  "language": {
    "name": "English (US)",
    "code": "en-US",
    "detectedLanguage": {
      "name": "Tagalog",
      "code": "tl-PH",
      "confidence": 0.165
    }
  },
  "matches": [
    {
      "message": "Possible spelling mistake found.",
      "shortMessage": "Spelling mistake",
      "replacements": [
        {
          "value": "birds"
        },
        {
          "value": "Burns"
        },
        {
          "value": "burns"
        },
        {
          "value": "buds"
        },
        {
          "value": "Burks"
        },
        {
          "value": "bards"
        },
        {
          "value": "curds"
        },
        {
          "value": "burrs"
        },
        {
          "value": "burls"
        },
        {
          "value": "burps"
        },
        {
          "value": "burs"
        },
        {
          "value": "burgs"
        },
        {
          "value": "burbs"
        },
        {
          "value": "turds"
        },
        {
          "value": "Kurds"
        }
      ],
      "offset": 0,
      "length": 5,
      "context": {
        "text": "burds",
        "offset": 0,
        "length": 5
      },
      "sentence": "burds",
      "type": {
        "typeName": "Other"
      },
      "rule": {
        "id": "MORFOLOGIK_RULE_EN_US",
        "description": "Possible spelling mistake",
        "issueType": "misspelling",
        "category": {
          "id": "TYPOS",
          "name": "Possible Typo"
        },
        "isPremium": false
      },
      "ignoreForIncompleteSentence": false,
      "contextForSureMatch": 0
    }
  ],
  "sentenceRanges": [
    [
      0,
      5
    ]
  ]
}
</pre>
</details>

It also allows you to set the langauge to "auto" and it tries to auto-detect. I'm not sure we should try to do that necessarily because queries probably err on the shorter side and likely wouldn't have enough context to guess the language from.

I have no idea how difficult it is to deploy either.

I would probably prefer `hunspell` over `languagetool`, mostly because I'm assuming it's easier to deploy `hunspell` into our Python app's container than it is to deploy an entire new web app micro-service.

**Comment by krysal** [comment]
I'd bet Elasticsearch already does that but we need to check of course.

**Comment by sarayourfriend** [comment]
It looks like if you're willing to put an entire dictionary of words into ES then yes, it can do some suggestion stuff: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-suggesters.html

I guess we don't even have to load up the whole dictionary, we can just rely on the existing indexed documents to source for the suggester. We'd miss "spell checking" words that don't have results, but those shouldn't matter in the end anyway :thinking: I wonder what the performance characteristics of it are vs hunspell or languagetool.

**Comment by raghuvar-arora-au2** [comment]
@sarayourfriend  I believe, it largely depends upon the type of service that you have in mind. Do we wanna limit the spell-check to 1) standard words, OR 2) everything that is indexed in ES (that inlcudes things like names)? 
if we are only looking to spell-check standard words, then spinning-up a microservice does not even make sense, while there are easier ways (hunspell) out there to get it done. 
The service would be immensly powerfull if we decide to go for 2), however I cannot find anything on the performance of suggest-API of ES.

**Comment by sarayourfriend** [comment]
Given the localization issues, would spell-checking against the index yield better results? I think it would, even if we didn't catch _every_ misspelling. The experience when you have no results is sad whether a misspelling exists or not. I don't think that it would necessarily improve the experience for a user if they searched for a slightly misspelled term with zero results, then get a suggestion for the corrected spelling and still no results come back.

What do you think? To me I think it makes sense to use the ES feature for this reason and because it requires the least amount of infrastructural changes (no new binaries or libraries or auxiliary services).

I'm wondering how this would work with localized searches as well.

@obulat What do you think about this? You worked a lot on the i18n stuff and also have been learning a lot about ES recently. I remember you sharing something at some point about localized queries as well :thinking: Do we need to keep that in mind as we look into this?

**Comment by raghuvar-arora-au2** [comment]
Yeah, using ES suggest-API should be pretty startightforward if we decide to ignore localization.
